src/control/family.lean

```diff
@@ -0,0 +1,666 @@
+  Π (f : drop α ⟶ drop β) (g : last α ⟶ last β), α ⟶ β
+| f g (sum.inl i) x := f x
+| f g (sum.inr i) x := g x
+
+/-- combine two arrows over different categories of families into an arrow on their product -/
+def append_fun {α β : fam I} {α' β' : fam J} (f : α ⟶ β) (g : α' ⟶ β') : (α.append1 α' ⟶ β.append1 β') :=
+split_fun f g
+
+lemma split_fun_comp {α β γ : fam (I⊕J)}
+  (f : drop α ⟶ drop β) (g : drop β ⟶ drop γ) (f' : last α ⟶ last β) (g' : last β ⟶ last γ) :
+  split_fun (f ≫ g) (f' ≫ g') = split_fun f f' ≫ split_fun g g' :=
+by ext (x|x) : 1; ext; refl
+
+lemma append_fun_comp {α β γ : fam I} {α' β' γ' : fam J}
+  (f : α ⟶ β) (f' : α' ⟶ β') (g : β ⟶ γ) (g' : β' ⟶ γ') :
+  append_fun (f ≫ g) (f' ≫ g') = append_fun f f' ≫ append_fun g g' :=
+by erw ← split_fun_comp; refl
+
+lemma append_fun_comp_right {α γ : fam I} {α' β' γ' : fam J}
+  (f : α ⟶ γ) (f' : α' ⟶ β') (g' : β' ⟶ γ') :
+  append_fun f (f' ≫ g') = append_fun f f' ≫ append_fun (𝟙 _) g' :=
+by erw ← split_fun_comp; refl
+
+lemma split_fun_comp_right {α : fam (I⊕J)} {β γ : fam J} {γ' : fam I}
+  (f : drop α ⟶ γ')
+  (f' : last α ⟶ β) (g' : β ⟶ γ) :
+  (split_fun f (f' ≫ g') : α ⟶ γ'.append1 γ) =
+  (split_fun f f' : α ⟶ γ'.append1 β) ≫ split_fun (𝟙 _) g' :=
+by rw [← split_fun_comp,category.comp_id]
+
+/-- take the left side of an arrow on a product of two families -/
+def drop_fun {α β : fam (I⊕J)} : Π (f : α ⟶ β), drop α ⟶ drop β
+| f i x := f x
+
+/-- take the right side of an arrow on a product of two families -/
+def last_fun {α β : fam (I⊕J)} : Π (f : α ⟶ β), last α ⟶ last β
+| f i x := f x
+
+theorem eq_of_drop_last_eq {α β : fam (I⊕J)} {f g : α ⟶ β}
+  (h₀ : ∀ j (x : α (sum.inl j)), drop_fun f x = drop_fun g x) (h₁ : last_fun f = last_fun g) :
+  f = g :=
+by { ext1 (i|j); ext1 x, apply h₀, apply congr_fun (congr_fun h₁ j), }
+-- by ext1 i; rcases i with ⟨j, ieq⟩ | ieq; [apply h₀, apply h₁]
+
+@[simp]
+theorem split_drop_fun_last_fun {α α' : fam (I⊕J)} (f : α ⟶ α') :
+  split_fun (drop_fun f) (last_fun f) = f :=
+eq_of_drop_last_eq (λ _ _, rfl) (funext $ λ _, funext $ λ _, rfl)
+
+theorem append_fun_id_id {α : fam I} {β : fam J} :
+  append_fun (𝟙 α) (𝟙 β) = 𝟙 _ :=
+by apply eq_of_drop_last_eq; intros; try { ext }; refl
+
+/-- `unit i` is an object in `fam I` such that has only one
+member and that its index is `i` -/
+inductive unit (i : I) : I → Type u
+| rfl {} : unit i
+
+instance {i : I} : inhabited (unit i i) := ⟨ unit.rfl ⟩
+
+/-- given a value of a type family, give an arrow to that object -/
+def value (i) (X : fam I) : X i → (unit i ⟶ X)
+| x j unit.rfl := x
+
+theorem value.get {i} {X : fam I} (f g : unit i ⟶ X) (h : f = g) : f unit.rfl = g unit.rfl :=
+by rw h
+
+theorem value.ext {i} {X : fam I} (f g : unit i ⟶ X) (h : f unit.rfl = g unit.rfl) : f = g :=
+by ext _ ⟨ ⟩; exact h
+
+@[simp]
+lemma value_eq  (i) (X : fam I) (x : X i) : Π {u : unit i i}, value i X x u = x
+| unit.rfl := rfl
+
+section subtype
+
+variables {F : fam I ⥤ fam J}
+
```

Comment:
It seems this isn't used until after you've done `Pred`?

Reply:
I think you're right. I can move that.